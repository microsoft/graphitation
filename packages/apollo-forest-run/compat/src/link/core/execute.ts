import { ApolloLink } from './ApolloLink';

export const execute = ApolloLink.execute;
