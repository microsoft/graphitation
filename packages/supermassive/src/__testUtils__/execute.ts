import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  OperationDefinitionNode,
  parse,
  ExecutionArgs,
  ExecutionArgs as GraphQLExecutionArgs,
  ExecutionResult as GraphQLExecutionResult,
} from "graphql";
import { executeWithSchema } from "../executeWithSchema";
import { typeDefs } from "../benchmarks/swapi-schema";
import resolvers from "../benchmarks/swapi-schema/resolvers";
import { ExecutionResult, UserResolvers } from "../types";
import models from "../benchmarks/swapi-schema/models";
import { executeWithoutSchema } from "../executeWithoutSchema";
import { extractMinimalViableSchemaForRequestDocument } from "../utilities/extractMinimalViableSchemaForRequestDocument";
import { PromiseOrValue } from "graphql/jsutils/PromiseOrValue";
import { ObjMap } from "../jsutils/ObjMap";
import { forAwaitEach, isAsyncIterable } from "iterall";

type GraphQLResult<TData = ObjMap<unknown>, TExtensions = ObjMap<unknown>> =
  | GraphQLExecutionResult<TData, TExtensions>
  | AsyncGenerator<GraphQLExecutionResult<TData, TExtensions>, void, void>;

export function createExecutionUtils(
  graphqlExecute: (args: ExecutionArgs) => PromiseOrValue<ExecutionResult>,
  graphqlSubscribe: (
    args: ExecutionArgs,
  ) => PromiseOrValue<AsyncIterableIterator<ExecutionResult> | ExecutionResult>,
) {
  async function compareResultsForExecuteWithSchema(
    schema: GraphQLSchema,
    query: string,
    variables?: Record<string, unknown>,
  ) {
    expect.assertions(1);
    const document = parse(query);
    const result = await drainExecution(
      await executeWithSchema({
        document,
        definitions: typeDefs,
        resolvers: resolvers as UserResolvers<unknown, unknown>,
        contextValue: {
          models,
        },
        variableValues: variables,
      }),
    );
    const validResult = await drainExecution(
      await graphqlExecuteOrSubscribe({
        document,
        contextValue: {
          models,
        },
        schema,
        variableValues: variables,
      }),
    );
    expect(result).toEqual(validResult);
  }

  async function compareResultForExecuteWithoutSchemaWithMVSAnnotation(
    schema: GraphQLSchema,
    query: string,
    variables: Record<string, unknown> = {},
  ) {
    expect.assertions(1);
    const document = parse(query);
    const { definitions } = extractMinimalViableSchemaForRequestDocument(
      schema,
      document,
    );
    const result = await drainExecution(
      await executeWithoutSchema({
        document,
        contextValue: {
          models,
        },
        schemaFragment: {
          schemaId: "test",
          definitions,
          resolvers: resolvers as UserResolvers,
        },
        variableValues: variables,
      }),
    );
    const validResult = await drainExecution(
      await graphqlExecuteOrSubscribe({
        document,
        contextValue: {
          models,
        },
        schema,
        variableValues: variables,
      }),
    );
    expect(result).toEqual(validResult);
  }

  function graphqlExecuteOrSubscribe(
    args: GraphQLExecutionArgs,
  ): PromiseOrValue<GraphQLResult> {
    const operationName = args.operationName;
    let operation: OperationDefinitionNode | undefined;
    for (const definition of (args.document as unknown as DocumentNode)
      .definitions) {
      switch (definition.kind) {
        case Kind.OPERATION_DEFINITION:
          if (operationName == null) {
            if (operation !== undefined) {
              throw new Error("Bad operation in test");
            }
            operation = definition;
          } else if (definition.name?.value === operationName) {
            operation = definition;
          }
          break;
      }
    }
    if (!operation) {
      throw new Error("Bad operation in test");
    }

    if (operation.operation === "subscription") {
      return graphqlSubscribe(args) as GraphQLResult;
    } else {
      return graphqlExecute(args) as GraphQLResult;
    }
  }

  async function drainExecution(
    result: ExecutionResult | GraphQLResult,
  ): Promise<unknown> {
    let processedResult;
    if (isAsyncIterable(result)) {
      processedResult = await drainAsyncGeneratorToArray(result);
    } else if ("subsequentResults" in result) {
      processedResult = {
        ...result,
        subsequentResults: await drainAsyncGeneratorToArray(
          result.subsequentResults,
        ),
      };
    } else {
      processedResult = result;
    }
    return processedResult;
  }

  async function drainAsyncGeneratorToArray<T>(
    collection: AsyncGenerator<T, void, void>,
  ): Promise<T[]> {
    const result: T[] = [];
    await forAwaitEach(collection, (item) => result.push(item));
    return result;
  }

  return {
    compareResultForExecuteWithoutSchemaWithMVSAnnotation,
    compareResultsForExecuteWithSchema,
    drainExecution,
    graphqlExecuteOrSubscribe,
  };
}
