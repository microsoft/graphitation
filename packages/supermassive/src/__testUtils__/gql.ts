import { DocumentNode, parse } from "graphql";

export type GqlResult = { raw: string; document: DocumentNode };

export function gql(raw: TemplateStringsArray): GqlResult {
  return { raw: raw[0], document: parse(raw[0]) };
}
