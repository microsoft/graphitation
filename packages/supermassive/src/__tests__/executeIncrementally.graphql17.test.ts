import {
  parse,
  GraphQLSchema,
  experimentalExecuteIncrementally as graphQLExecute,
  experimentalSubscribeIncrementally as graphQLSubscribe,
} from "graphql";
import { makeSchema } from "../benchmarks/swapi-schema";
import models from "../benchmarks/swapi-schema/models";
import { createExecutionUtils } from "../__testUtils__/execute";

const {
  compareResultsForExecuteWithSchema,
  compareResultForExecuteWithoutSchemaWithMVSAnnotation,
  drainExecution,
  graphqlExecuteOrSubscribe,
} = createExecutionUtils(graphQLExecute, graphQLSubscribe);

interface TestCase {
  name: string;
  document: string;
  variables?: Record<string, unknown>;
}

const testCases: Array<TestCase> = [
  {
    name: "benchmark kitchen-sink query",
    document: `
    {
      empireHero: person(id: 1) {
        ...PersonFields
        height
        mass
      }
    
      jediHero: person(id: 4) {
        ...PersonFields
      }
    }
    
    fragment PersonFields on Person {
      name
      gender
      birth_year
      # This is async iterator
      starships {
        name
      }
    }  
    `,
  },
  // defer
  {
    name: "@defer on inline fragment",
    document: `
    {
      person(id: 1) {
        name
        ... on Person @defer {
          gender
          birth_year
        }
      }
    }
    `,
  },
  {
    name: "@defer on fragment",
    document: `
    {
      person(id: 1) {
        name
        ...DeferredPerson @defer
      }
    }

    fragment DeferredPerson on Person {
      gender
      birth_year
    }
    `,
  },
  {
    name: "@defer multiple fragments",
    document: `
    {
      person(id: 1) {
        name
        ... on Person @defer {
          gender
        }
        ...DeferredPerson @defer
      }
    }

    fragment DeferredPerson on Person {
      birth_year
    }
    `,
  },
  {
    name: "@defer overlapping fragment",
    document: `
    {
      person(id: 1) {
        name
        ... on Person @defer {
          gender
        }
        ...DeferredPerson @defer
      }
    }

    fragment DeferredPerson on Person {
      gender
      birth_year
    }
    `,
  },

  {
    name: "@defer nested",
    document: `
    {
      person(id: 1) {
        name
        ...DeferredPerson @defer
      }
    }

    fragment DeferredPerson on Person {
      birth_year
      ... on Person @defer {
        films {
          title
        }
      }
    }
    `,
  },

  {
    name: "@defer label simple",
    document: `
   {
      person(id: 1) {
        name
        ... on Person @defer(label: "INLINE") {
          gender
        }
        ...DeferredPerson @defer(label: null)
      }
    }

    fragment DeferredPerson on Person {
      birth_year
    }
    `,
  },

  {
    name: "@defer on list",
    document: `
    {
      person(id: 1) {
        name
        films {
          ... on Film @defer {
            title
          }
        }
      }
    }
    `,
  },

  {
    name: "@defer on query with union",
    document: `
    {
      search(search: "Lu") {
        __typename
        ...on Person @defer {
          name
        }
        ...on Planet {
          name
        }
        ... on Transport {
          name
        }
        ... on Species {
          name
        }
        ... on Vehicle{
          name
        }
      }
    }
    `,
  },

  // TODO: Does not error in graphql-js even though it should in spec
  // {
  //   name: "@defer label as  variable error",
  //   document: `
  //   query ($label: String) {
  //     person(id: 1) {
  //       name
  //       ... on Person @defer(label: $label) {
  //         gender
  //       }
  //       ...DeferredPerson @defer
  //     }
  //   }

  //   fragment DeferredPerson on Person {
  //     birth_year
  //   }
  //   `,
  //   variables: {
  //     inlineLabel: "INLINE",
  //   },
  // },

  /// TODO: Does not error in graphql-js even though it should by spec
  // {
  //   name: "@defer label duplicate labels",
  //   document: `
  //   query  {
  //     person(id: 1) {
  //       name
  //       ... on Person @defer(label: "SAME") {
  //         gender
  //       }
  //       ...DeferredPerson @defer(label: "SAME")
  //     }
  //   }

  //   fragment DeferredPerson on Person {
  //     birth_year
  //   }
  //   `,
  // },

  {
    name: "@defer if",
    document: `
   {
      person(id: 1) {
        name
        ... on Person @defer(if: true) {
          gender
        }
        ...DeferredPerson @defer(if: false)
      }
    }

    fragment DeferredPerson on Person {
      birth_year
    }
    `,
  },

  {
    name: "@defer crossselection",
    document: `
   {
      person(id: 1) {
        name
        gender
        ... on Person @defer {
          gender
        }
      }
    }
    `,
  },

  {
    name: "@defer skip/include",
    document: `
   {
      person(id: 1) {
        name
        ... on Person @defer {
          gender @skip(if: true)
        }
        ... on Person @defer {
          birth_year @include(if: false)
        }
        ... on Person @defer @skip(if: true) {
          skin_color
        }
        ... on Person @defer @include(if: false) {
          hair_color
        }
      }
    }
    `,
  },

  {
    name: "@stream basic",
    document: `
    {
      person(id: 1) {
        name
        films @stream {
          title
        }
      }
    }
    `,
  },

  {
    name: "@stream inside @defer",
    document: `
    {
      person(id: 1) {
        name
        ...DeferredPerson @defer
      }
    }

    fragment DeferredPerson on Person {
      birth_year
      ... on Person @defer {
        films @stream {
          title
        }
      }
    }`,
  },

  {
    name: "@stream if",
    document: `
    {
      person(id: 1) {
        name
        films @stream(if: true) {
          title
        }
        starships @stream(if: false) {
          name
        }
      }
    }
    `,
  },

  {
    name: "@stream multiple",
    document: `
      {
        person(id: 1) {
          name
          films @stream {
            title
          }
          starships @stream {
            name
          }
        }
      }
      `,
  },

  {
    name: "@stream nested",
    document: `
      {
        person(id: 1) {
          name
          films @stream {
            title
            planets @stream
          }
        }
      }
      `,
  },

  {
    name: "@stream label",
    document: `
      {
        person(id: 1) {
          name
          films @stream(label: "FILMS") {
            title
          }
          starships @stream(label: "STARSHIPS") {
            name
          }
        }
      }
      `,
  },
  // TODO: not throwing in graphql-js but should by spec - duplicate labels, labels as vars

  {
    name: "@stream initialCount",
    document: `
    {
      person(id: 1) {
        name
        films @stream(label: "FILMS", initialCount: 2) {
          title
        }
      }
    }
    `,
  },

  {
    name: "@stream/defer errors",
    document: `
      {
        person(id: 1) {
          name
          ... on Person @defer {
            bubblingError
          }
          bubblingListError @stream
        }
      }`,
  },
];

describe("graphql-js snapshot check to ensure test stability", () => {
  let schema: GraphQLSchema;
  beforeEach(() => {
    jest.resetAllMocks();
    schema = makeSchema();
  });

  test.each(testCases)("$name", async ({ document, variables }: TestCase) => {
    expect.assertions(1);
    const parsedDocument = parse(document);
    const result = await drainExecution(
      await graphqlExecuteOrSubscribe({
        schema,
        document: parsedDocument,
        contextValue: {
          models,
        },
        variableValues: variables,
      }),
    );
    expect(result).toMatchSnapshot();
  });
});

describe("executeWithSchema", () => {
  let schema: GraphQLSchema;
  beforeEach(() => {
    jest.resetAllMocks();
    schema = makeSchema();
  });

  test.each(testCases)("$name", async ({ document, variables }: TestCase) => {
    await compareResultsForExecuteWithSchema(schema, document, variables);
  });
});

describe("executeWithoutSchema - minimal viable schema annotation", () => {
  let schema: GraphQLSchema;
  beforeEach(() => {
    jest.resetAllMocks();
    schema = makeSchema();
  });
  test.each(testCases)("$name", async ({ document, variables }: TestCase) => {
    await compareResultForExecuteWithoutSchemaWithMVSAnnotation(
      schema,
      document,
      variables,
    );
  });
});
