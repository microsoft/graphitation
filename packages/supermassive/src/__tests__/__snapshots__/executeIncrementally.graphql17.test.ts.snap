// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-js snapshot check to ensure test stability @defer crossselection 1`] = `
{
  "data": {
    "person": {
      "gender": "male",
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer if 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "birth_year": "19BBY",
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "gender": "male",
          },
          "id": "0",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer label simple 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "label": "INLINE",
        "path": [
          "person",
        ],
      },
      {
        "id": "1",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
        {
          "id": "1",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "gender": "male",
          },
          "id": "0",
        },
        {
          "data": {
            "birth_year": "19BBY",
          },
          "id": "1",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer multiple fragments 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
      {
        "id": "1",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
        {
          "id": "1",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "gender": "male",
          },
          "id": "0",
        },
        {
          "data": {
            "birth_year": "19BBY",
          },
          "id": "1",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer nested 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": true,
      "incremental": [
        {
          "data": {
            "birth_year": "19BBY",
          },
          "id": "0",
        },
      ],
      "pending": [
        {
          "id": "1",
          "path": [
            "person",
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "1",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "films": [
              {
                "title": "A New Hope",
              },
              {
                "title": "The Empire Strikes Back",
              },
              {
                "title": "Return of the Jedi",
              },
              {
                "title": "Revenge of the Sith",
              },
            ],
          },
          "id": "1",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer on fragment 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "birth_year": "19BBY",
            "gender": "male",
          },
          "id": "0",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer on inline fragment 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "birth_year": "19BBY",
            "gender": "male",
          },
          "id": "0",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer on list 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "films": [
          {},
          {},
          {},
          {},
        ],
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
          "films",
          0,
        ],
      },
      {
        "id": "1",
        "path": [
          "person",
          "films",
          1,
        ],
      },
      {
        "id": "2",
        "path": [
          "person",
          "films",
          2,
        ],
      },
      {
        "id": "3",
        "path": [
          "person",
          "films",
          3,
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
        {
          "id": "1",
        },
        {
          "id": "2",
        },
        {
          "id": "3",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "title": "A New Hope",
          },
          "id": "0",
        },
        {
          "data": {
            "title": "The Empire Strikes Back",
          },
          "id": "1",
        },
        {
          "data": {
            "title": "Return of the Jedi",
          },
          "id": "2",
        },
        {
          "data": {
            "title": "Revenge of the Sith",
          },
          "id": "3",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer on query with union 1`] = `
{
  "initialResult": {
    "data": {
      "search": [
        {
          "__typename": "Person",
        },
        {
          "__typename": "Person",
        },
        {
          "__typename": "Planet",
          "name": "Felucia",
        },
        {
          "__typename": "Planet",
          "name": "Sullust",
        },
        {
          "__typename": "Species",
          "name": "Sullustan",
        },
        {
          "__typename": "Transport",
          "name": "Raddaugh Gnasp fluttercraft",
        },
        {
          "__typename": "Vehicle",
          "name": "Raddaugh Gnasp fluttercraft",
        },
      ],
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "search",
          0,
        ],
      },
      {
        "id": "1",
        "path": [
          "search",
          1,
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
        {
          "id": "1",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "name": "Luke Skywalker",
          },
          "id": "0",
        },
        {
          "data": {
            "name": "Luminara Unduli",
          },
          "id": "1",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer overlapping fragment 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
      {
        "id": "1",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
        {
          "id": "1",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "gender": "male",
          },
          "id": "0",
        },
        {
          "data": {
            "birth_year": "19BBY",
          },
          "id": "1",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @defer skip/include 1`] = `
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream basic 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "films": [],
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
          "films",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "A New Hope",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "The Empire Strikes Back",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Return of the Jedi",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Revenge of the Sith",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream if 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "films": [],
        "name": "Luke Skywalker",
        "starships": [
          {
            "name": "X-wing",
          },
          {
            "name": "Imperial shuttle",
          },
        ],
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
          "films",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "A New Hope",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "The Empire Strikes Back",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Return of the Jedi",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Revenge of the Sith",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream initialCount 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "films": [
          {
            "title": "A New Hope",
          },
          {
            "title": "The Empire Strikes Back",
          },
        ],
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "label": "FILMS",
        "path": [
          "person",
          "films",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Return of the Jedi",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Revenge of the Sith",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream inside @defer 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": true,
      "incremental": [
        {
          "data": {
            "birth_year": "19BBY",
          },
          "id": "0",
        },
      ],
      "pending": [
        {
          "id": "1",
          "path": [
            "person",
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "1",
        },
      ],
      "hasNext": true,
      "incremental": [
        {
          "data": {
            "films": [],
          },
          "id": "1",
        },
      ],
      "pending": [
        {
          "id": "2",
          "path": [
            "person",
            "films",
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "2",
          "items": [
            {
              "title": "A New Hope",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "2",
          "items": [
            {
              "title": "The Empire Strikes Back",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "2",
          "items": [
            {
              "title": "Return of the Jedi",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "2",
          "items": [
            {
              "title": "Revenge of the Sith",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "2",
        },
      ],
      "hasNext": false,
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream label 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "films": [],
        "name": "Luke Skywalker",
        "starships": [],
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "label": "FILMS",
        "path": [
          "person",
          "films",
        ],
      },
      {
        "id": "1",
        "label": "STARSHIPS",
        "path": [
          "person",
          "starships",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "A New Hope",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "1",
          "items": [
            {
              "name": "X-wing",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "The Empire Strikes Back",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "1",
          "items": [
            {
              "name": "Imperial shuttle",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "1",
        },
      ],
      "hasNext": true,
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Return of the Jedi",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Revenge of the Sith",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream multiple 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "films": [],
        "name": "Luke Skywalker",
        "starships": [],
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
          "films",
        ],
      },
      {
        "id": "1",
        "path": [
          "person",
          "starships",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "A New Hope",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "1",
          "items": [
            {
              "name": "X-wing",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "The Empire Strikes Back",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "1",
          "items": [
            {
              "name": "Imperial shuttle",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "1",
        },
      ],
      "hasNext": true,
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Return of the Jedi",
            },
          ],
        },
      ],
    },
    {
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "title": "Revenge of the Sith",
            },
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream nested 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "films": [],
        "name": "Luke Skywalker",
      },
    },
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
          "films",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "1",
        },
      ],
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "planets": [],
              "title": "A New Hope",
            },
          ],
        },
        {
          "id": "1",
          "items": [
            {},
            {},
            {},
          ],
        },
      ],
      "pending": [
        {
          "id": "1",
          "path": [
            "person",
            "films",
            0,
            "planets",
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "2",
        },
      ],
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "planets": [],
              "title": "The Empire Strikes Back",
            },
          ],
        },
        {
          "id": "2",
          "items": [
            {},
            {},
            {},
            {},
          ],
        },
      ],
      "pending": [
        {
          "id": "2",
          "path": [
            "person",
            "films",
            1,
            "planets",
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "3",
        },
      ],
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "planets": [],
              "title": "Return of the Jedi",
            },
          ],
        },
        {
          "id": "3",
          "items": [
            {},
            {},
            {},
            {},
            {},
          ],
        },
      ],
      "pending": [
        {
          "id": "3",
          "path": [
            "person",
            "films",
            2,
            "planets",
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "4",
        },
      ],
      "hasNext": true,
      "incremental": [
        {
          "id": "0",
          "items": [
            {
              "planets": [],
              "title": "Revenge of the Sith",
            },
          ],
        },
        {
          "id": "4",
          "items": [
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
          ],
        },
      ],
      "pending": [
        {
          "id": "4",
          "path": [
            "person",
            "films",
            3,
            "planets",
          ],
        },
      ],
    },
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability @stream/defer errors 1`] = `
{
  "initialResult": {
    "data": {
      "person": {
        "bubblingListError": null,
        "name": "Luke Skywalker",
      },
    },
    "errors": [
      [GraphQLError: Bubbling in list!],
    ],
    "hasNext": true,
    "pending": [
      {
        "id": "0",
        "path": [
          "person",
        ],
      },
    ],
  },
  "subsequentResults": [
    {
      "completed": [
        {
          "id": "0",
        },
      ],
      "hasNext": false,
      "incremental": [
        {
          "data": {
            "bubblingError": null,
          },
          "errors": [
            [GraphQLError: Bubbling!],
          ],
          "id": "0",
        },
      ],
    },
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability benchmark kitchen-sink query 1`] = `
{
  "data": {
    "empireHero": {
      "birth_year": "19BBY",
      "gender": "male",
      "height": 172,
      "mass": 77,
      "name": "Luke Skywalker",
      "starships": [
        {
          "name": "X-wing",
        },
        {
          "name": "Imperial shuttle",
        },
      ],
    },
    "jediHero": {
      "birth_year": "41.9BBY",
      "gender": "male",
      "name": "Darth Vader",
      "starships": [
        {
          "name": "TIE Advanced x1",
        },
      ],
    },
  },
}
`;
